### DISCUSSION TITLE
New Renderer (v4)

### DISCUSSION THREAD

**Discussion by Gazook89** [naturalcrit/homebrewery#2878] [Ideas]
Today there was a little discussion about a new approach to rendering documents, namely utilizing d3.js in conjunction with Codemirror to output an HTML canvas rather than styled DOM.  Please correct me if I misunderstood that.

It might not ever happen, and we can't be sure what the final outcome is, but it's good to drop that info in one spot.  Any change to the renderer is likely to be delayed to a "version 4" since we don't want to screw up existing brews.  

Benefits of something like d3.js:
- Better interaction in both directions between Editor and Renderer, enabling such things as drag'n'drop movement of elements, and style widgets visible in the renderer that update text in the editor.
- More options for creating exports:  HTML Canvas to JPG or PNG files.

Cons:
- Not sure what HTML Canvas does for accessibility?  
- Very large undertaking
- Additional level of complexity for developing, not to mention needing to maintain the Canvas system + the existing system.

**Comment by T3sT3ro** [comment]
- I'm not sure how theming would be approached in canvas. With the DOM approach it is very simple and customizable.
- As for accessibility, I remember that canvas is not accessible to screen readers, and any rewrite to canvas could be a big undertaking. How d3.js approaches that, I don't know. 
- I also am not sure how canvas would play out with exporting to PDFs.
- How would selection work on a canvas?
- Maybe also another thing related to the renderer — extract the renderer to a separate package that can be utilized standalone in other projects.
- [Is this slideshow from 2014 still relevant today?](https://www.w3.org/Talks/2014/0510-canvas-a11y/#1)

I don't know, I have a feeling a canvas renderer could be a bad idea.

I'm not sure how you approach considering using canvas or 3d.js in your project. If I were to decide for or against it, I would start with "what immediate existing problems and missing features would they solve" instead of "By using those techs, what can we do in the project". If a simple WYSWIG editor and drag and drop are the only core features, then maybe it's an overkill?

**Comment by 5e-Cleric** [comment]
I don't think Canvas is the way to go forward, specially with so much work that has gone into the DOM for links and variables, which would have to be redone.

We can still use this discussion as a place to talk about what features we might want V4 to have.

**Reply by dbolack-ab** [reply]
I'm not even certain that using a canvas driven setup would work due to canvas tainting.

**Comment by 5e-Cleric** [comment]
Closing this in favor of #2879
